jvb/src/main/kotlin/org/jitsi/videobridge/relay/Relay.kt

```diff
@@ -0,0 +1,629 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.videobridge.relay
+
+import org.jitsi.nlj.Features
+import org.jitsi.nlj.MediaSourceDesc
+import org.jitsi.nlj.PacketHandler
+import org.jitsi.nlj.PacketInfo
+import org.jitsi.nlj.Transceiver
+import org.jitsi.nlj.TransceiverEventHandler
+import org.jitsi.nlj.rtp.SsrcAssociationType
+import org.jitsi.nlj.srtp.TlsRole
+import org.jitsi.nlj.stats.EndpointConnectionStats
+import org.jitsi.nlj.stats.PacketDelayStats
+import org.jitsi.nlj.transform.node.ConsumerNode
+import org.jitsi.nlj.util.Bandwidth
+import org.jitsi.nlj.util.LocalSsrcAssociation
+import org.jitsi.nlj.util.PacketInfoQueue
+import org.jitsi.nlj.util.RemoteSsrcAssociation
+import org.jitsi.rtp.Packet
+import org.jitsi.rtp.UnparsedPacket
+import org.jitsi.rtp.extensions.looksLikeRtcp
+import org.jitsi.rtp.extensions.looksLikeRtp
+import org.jitsi.rtp.rtp.RtpPacket
+import org.jitsi.utils.MediaType
+import org.jitsi.utils.OrderedJsonObject
+import org.jitsi.utils.event.EventEmitter
+import org.jitsi.utils.event.SyncEventEmitter
+import org.jitsi.utils.logging2.Logger
+import org.jitsi.utils.logging2.cdebug
+import org.jitsi.utils.logging2.createChildLogger
+import org.jitsi.utils.queue.CountingErrorHandler
+import org.jitsi.videobridge.AbstractEndpoint
+import org.jitsi.videobridge.Conference
+import org.jitsi.videobridge.EncodingsManager
+import org.jitsi.videobridge.Endpoint
+import org.jitsi.videobridge.PotentialPacketHandler
+import org.jitsi.videobridge.TransportConfig
+import org.jitsi.videobridge.message.BridgeChannelMessage
+import org.jitsi.videobridge.octo.OctoPacketInfo
+import org.jitsi.videobridge.rest.root.debug.EndpointDebugFeatures
+import org.jitsi.videobridge.transport.dtls.DtlsTransport
+import org.jitsi.videobridge.transport.ice.IceTransport
+import org.jitsi.videobridge.util.ByteBufferPool
+import org.jitsi.videobridge.util.TaskPools
+import org.jitsi.videobridge.util.looksLikeDtls
+import org.jitsi.videobridge.websocket.colibriWebSocketServiceSupplier
+import org.jitsi.xmpp.extensions.colibri.WebSocketPacketExtension
+import org.jitsi.xmpp.extensions.jingle.DtlsFingerprintPacketExtension
+import org.jitsi.xmpp.extensions.jingle.IceUdpTransportPacketExtension
+import org.json.simple.JSONObject
+import java.time.Clock
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+import java.util.function.Supplier
+
+/**
+ * Models a relay (remote videobridge) in a [Conference].
+ */
+/* TODO: figure out how best to share code between this and [Endpoint], without multiple inheritance. */
+class Relay @JvmOverloads constructor(
```

Comment:
`Relay` models a collection of remote endpoints, is that right?  It seems like it having a `conference` member kinda inverts the hierarchy a bit: ideally the methods it needs to call there would be passed here by some kind of interface rather than making this depend directly on `Conference`.

Reply:
It's a collection of remote endpoints, but also the transport connection to it.

I don't really see that this is that different from `AbstractEndpoint` having a `conference` member.  It might be good to abstract all of that, but I don't think it'd be sensible to abstract it just for relays.

Reply:
Ah, I forgot `AbstractEndpoint` had `conference` member, yeah I agree would be nice if that was gone too, but not so bad that this doesn't do anything too different, I guess.